src/audio/dai-zephyr.c

```diff
@@ -0,0 +1,1408 @@
+	    copy_bytes + avail_bytes < dd->period_bytes)
+		comp_warn(dev, "dai_copy(): Copy_bytes %d + avail bytes %d < period bytes %d, possible glitch",
+			  copy_bytes, avail_bytes, dd->period_bytes);
+	else if (dev->direction == SOF_IPC_STREAM_CAPTURE &&
+		 copy_bytes + free_bytes < dd->period_bytes)
+		comp_warn(dev, "dai_copy(): Copy_bytes %d + free bytes %d < period bytes %d, possible glitch",
+			  copy_bytes, free_bytes, dd->period_bytes);
+
+	/* return if nothing to copy */
+	if (!copy_bytes) {
+		comp_warn(dev, "dai_copy(): nothing to copy");
+		return 0;
+	}
+
+	ret = dma_copy(dd->chan, copy_bytes, 0);
+	if (ret < 0) {
+		dai_report_xrun(dev, copy_bytes);
+		return ret;
+	}
+
+	dai_dma_position_update(dev);
+
+	return ret;
+}
+
+/**
+ * \brief Get DAI parameters and configure timestamping
+ * \param[in, out] dev DAI device.
+ * \return Error code.
+ *
+ * This function retrieves various DAI parameters such as type, direction, index, and DMA
+ * controller information those are needed when configuring HW timestamping. Note that
+ * DAI must be prepared before this function is used (for DMA information). If not, an error
+ * is returned.
+ */
+static int dai_ts_config_op(struct comp_dev *dev)
+{
+#if 0
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct timestamp_cfg *cfg = &dd->ts_config;
+	struct ipc_config_dai *dai = &dd->ipc_config;
+
+	comp_dbg(dev, "dai_ts_config()");
+	if (!dd->chan) {
+		comp_err(dev, "dai_ts_config(), No DMA channel information");
+		return -EINVAL;
+	}
+
+	cfg->type = dd->dai->drv->type;
+	cfg->direction = dai->direction;
+	cfg->index = dd->dai->index;
+	cfg->dma_id = dd->dma->plat_data.id;
+	cfg->dma_chan_index = dd->chan->index;
+	cfg->dma_chan_count = dd->dma->plat_data.channels;
+	if (!dd->dai->drv->ts_ops.ts_config)
+		return -ENXIO;
+
+	return dd->dai->drv->ts_ops.ts_config(dd->dai, cfg);
+#endif
+	return 0;
+}
+
+static int dai_ts_start_op(struct comp_dev *dev)
+{
+#if 0
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "dai_ts_start()");
+	if (!dd->dai->drv->ts_ops.ts_start)
+		return -ENXIO;
+
+	return dd->dai->drv->ts_ops.ts_start(dd->dai, &dd->ts_config);
+#endif
+	return 0;
+}
+
+static int dai_ts_get_op(struct comp_dev *dev, struct timestamp_data *tsd)
+{
+#if 0
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "dai_ts_get()");
+	if (!dd->dai->drv->ts_ops.ts_get)
+		return -ENXIO;
+
+	return dd->dai->drv->ts_ops.ts_get(dd->dai, &dd->ts_config, tsd);
+#endif
+	return 0;
+}
+
+static int dai_ts_stop_op(struct comp_dev *dev)
+{
+#if 0
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "dai_ts_stop()");
+	if (!dd->dai->drv->ts_ops.ts_stop)
+		return -ENXIO;
+
+	return dd->dai->drv->ts_ops.ts_stop(dd->dai, &dd->ts_config);
+#endif
+	return 0;
+}
+
+uint32_t dai_get_init_delay_ms(struct dai *dai)
+{
+	const struct dai_properties *props = dai_get_properties(dai->drv, 0, 0);
+
+	return props->reg_init_delay;
+}
+
+struct dai_group_list {
```

Comment:
Lets move dai_group to another source and header for simplicity. It's probably mostly common code between xtos and zephyr.

Reply:
1 dai_group function registers notifier for dai's trigger internal, so this might be very difficult as this is not exported.